include/nlohmann/detail/output/serializer.hpp

```diff
@@ -104,272 +104,326 @@ class serializer
-
-                    o->write_character('\n');
-                    o->write_characters(indent_string.c_str(), current_indent);
-                    o->write_character('}');
-                }
-                else
-                {
-                    o->write_character('{');
-
-                    // first n-1 elements
-                    auto i = val.m_data.m_value.object->cbegin();
-                    for (std::size_t cnt = 0; cnt < val.m_data.m_value.object->size() - 1; ++cnt, ++i)
+                    if (pretty_print)
                     {
-                        o->write_character('\"');
-                        dump_escaped(i->first, ensure_ascii);
-                        o->write_characters("\":", 2);
-                        dump(i->second, false, ensure_ascii, indent_step, current_indent);
-                        o->write_character(',');
-                    }
+                        // variable to hold indentation for recursive calls
+                        unsigned int new_indent = current_indent;
+                        auto i = val.m_data.m_value.object->cbegin();
+                        if (offset == 0)
+                        {
+                            o->write_characters("{\n", 2);
 
-                    // last element
-                    JSON_ASSERT(i != val.m_data.m_value.object->cend());
-                    JSON_ASSERT(std::next(i) == val.m_data.m_value.object->cend());
-                    o->write_character('\"');
-                    dump_escaped(i->first, ensure_ascii);
-                    o->write_characters("\":", 2);
-                    dump(i->second, false, ensure_ascii, indent_step, current_indent);
+                            new_indent = current_indent + indent_step;
+                            if (JSON_HEDLEY_UNLIKELY(indent_string.size() < new_indent))
+                            {
+                                indent_string.resize(indent_string.size() * 2, ' ');
+                            }
+                        }
+                        else if (offset < val.m_data.m_value.object->size())
+                        {
+                            o->write_characters(",\n", 2);
+                            i = iterator;
+                        }
 
-                    o->write_character('}');
-                }
+                        // first n-1 elements
+                        if (offset < val.m_data.m_value.object->size() - 1)
+                        {
+                            o->write_characters(indent_string.c_str(), new_indent);
+                            o->write_character('\"');
+                            dump_escaped(i->first, ensure_ascii);
+                            o->write_characters("\": ", 3);
+                            stack.push_back(dump_state(val, new_indent, offset + 1, std::next(i)));
+                            stack.push_back(dump_state(i->second, new_indent, 0));
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/agdas/json/pull/2#issuecomment-2418484953>#868c9b</a><span> - 10/17/2024, 05:03 am</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The new implementation creates temporary dump_state objects which are then copied into the stack. This could lead to unnecessary object constructions and copies.
</div>
<div id="fix">
<b>Fix:</b> Consider using std::move when pushing dump_state objects onto the stack to avoid unnecessary copies. Also, you could use emplace_back instead of push_back for more efficient insertion.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -176,2 +176,2 @@
 -                            stack.push_back(dump_state(val, new_indent, offset + 1, std::next(i)));
 -                            stack.push_back(dump_state(i->second, new_indent, 0));
 +                            stack.emplace_back(val, new_indent, offset + 1, std::next(i));
 +                            stack.emplace_back(i->second, new_indent, 0);
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged